### ISSUE TITLE
OpCompositeConstruct causes driver crash in rare (but seemingly valid) case

### ISSUE THREAD

**Issue by beholdnec** [KhronosGroup/SPIRV-Headers#21]
I have a project that generates SPIR-V directly, rather than compiling it from GLSL via glslangValidator.

The code under consideration takes a vector of 3 floats (vec3) and creates a vector of 4 floats (vec4) by concatenating the vec3 with a fourth component, 1.0. The equivalent GLSL is:
```GLSL
vec3 v3 = input.xyz;
vec4 v4 = vec4(v3, 1.0); // <-- This is what I want to accomplish.
```

To accomplish this, I generated what seemed to be the most straightforward SPIR-V code:
```
%1 = OpLoad %v3float %...
%2 = OpConstant %float 1
%3 = OpCompositeConstruct %v4float %1 %2
```
However, when attempting to use this code in a vertex shader, it causes an access violation in the latest AMD driver (16.9.3).

When the same function is performed in GLSL and compiled by glslangValidator, it creates the vec4 a different way. It extracts each float component from the vec3 individually and constructs a vec4 from 4 floats. glslangValidator generates SPIR-V code like this:
```
%1 = OpLoad %v3float %...
%2 = OpCompositeExtract %float %1 0
%3 = OpCompositeExtract %float %1 1
%4 = OpCompositeExtract %float %1 2
%5 = OpConstant %float 1
%6 = OpCompositeConstruct %v4float %2 %3 %4 %5
```
This code works in the latest AMD driver.

Now, the SPIR-V spec seems to imply that the original code (the one that crashes) is valid and should do what I expect. OpCompositeConstruct should create a v4float if the input parameters are a v3float and a float. The SPIR-V 1.0 Specification says:

>OpCompositeConstruct

>Construct a new composite object from a set of constituent objects that will fully form it.

>Result Type must be a composite type, whose top-level members/elements/components/columns have the same type as the types of the operands, with one exception. ***The exception is that for constructing a vector, the operands may also be vectors with the same component type as the Result Type component type. When constructing a vector, the total number of components in all the operands must equal the number of components in Result Type.***

The "exception" allows combining vectors of different sizes into a new vector whose size is the total of all constituents. But since glslangValidator does not generate such code, I suspect that driver programmers have neglected to support this case.

In closing, I have two questions. Do I understand correctly that my original SPIR-V is valid? If so, where can I submit a bug report to AMD?

**Comment by johnkslang** [comment]
It seems your understanding of SPIR-V is correct.  The OpType stuff is missing in your example, so I don't know about driver response to it, but conceptually, it seems correct.

**Comment by beholdnec** [comment]
Here is a complete listing of the SPIR-V vertex shader that causes a crash:
```
; SPIR-V
; Version: 1.1
; Generator: Khronos; 0
; Bound: 15
; Schema: 0
               OpCapability Shader
               OpMemoryModel Logical Simple
               OpEntryPoint Vertex %1 "main" %2 %3
               OpDecorate %2 Location 0
               OpDecorate %3 BuiltIn Position
       %void = OpTypeVoid
      %float = OpTypeFloat 32
    %v3float = OpTypeVector %float 3
    %v4float = OpTypeVector %float 4
%_ptr_Input_v3float = OpTypePointer Input %v3float
%_ptr_Output_v4float = OpTypePointer Output %v4float
         %10 = OpTypeFunction %void
         %11 = OpConstant %float 1
          %2 = OpVariable %_ptr_Input_v3float Input
          %3 = OpVariable %_ptr_Output_v4float Output
          %1 = OpFunction %void None %10
         %12 = OpLabel
         %13 = OpLoad %v3float %2
         %14 = OpCompositeConstruct %v4float %13 %11
               OpStore %3 %14
               OpReturn
               OpFunctionEnd
```
My program throws the following error:
>Exception thrown at 0x0FB4FC54 (amdvlk32.dll) in VulkanTest.exe: 0xC0000005: Access violation reading location 0x00000018.

When I use OpCompositeExtract to pull each component individually, it works perfectly.

**Comment by amdrexu** [comment]
Your case is valid. vec4 = vec3 + float is valid according to SPIR-V and this issue is fixed in AMD driver recently. Future version will include the fix. Thank you.

**Comment by johnkslang** [comment]
Closing here, as this is not a glslang issue.

**Comment by beholdnec** [comment]
This bug is fixed in Catalyst 16.9.2. Thank you.

Fixed driver info for reference:
>        apiVersion     = 0x400015  (1.0.21)
>        driverVersion  = 4206592 (0x403000)
>        vendorID       = 0x1002
>        deviceID       = 0x67b1
>        deviceType     = DISCRETE_GPU
>        deviceName     = AMD Radeon (TM) R9 390 Series
